util/url-encoding/src/payment_request.rs

```diff
@@ -0,0 +1,345 @@
+    #[test]
+    fn example_fog_public_address() {
+        let identity = RootIdentity {
+            root_entropy: [0u8; 32],
+            fog_url: Some("fog://example.com".to_owned()),
+        };
+
+        let acct = AccountKey::from(&identity);
+
+        let addr = acct.default_subaddress();
+
+        let payload = PaymentRequest::from(&addr);
+
+        let mob_url = MobUrl::try_from(&payload)
+            .map_err(|err| {
+                panic!("Error when decoding payload {:?}: {}", payload, err);
+            })
+            .unwrap();
+
+        assert_eq!(mob_url.as_ref(), "mob://example.com/9i_xwzoihbGu5hLthygfLGi7K1sPFDmhPkq3KPmO-2p4kBwRg06ELfa-mMEnlTUT4RYJXUEizCfYB7RRHLgeEWfP?s=CQkJCfSo");
+
+        let payload2 = PaymentRequest::try_from(&mob_url).unwrap();
+
+        assert_eq!(payload, payload2);
+    }
+
+    // Test an example fogless public address being parsed to mob url
+    #[test]
+    fn example_fogless_public_address() {
+        let identity = RootIdentity {
+            root_entropy: [0u8; 32],
+            fog_url: None,
+        };
+
+        let acct = AccountKey::from(&identity);
+
+        let addr = acct.default_subaddress();
+
+        let payload = PaymentRequest::from(&addr);
+
+        let mob_url = MobUrl::try_from(&payload)
+            .map_err(|err| {
+                panic!("Error when decoding payload {:?}: {}", payload, err);
+            })
+            .unwrap();
+
+        assert_eq!(mob_url.as_ref(), "mob:///9i_xwzoihbGu5hLthygfLGi7K1sPFDmhPkq3KPmO-2p4kBwRg06ELfa-mMEnlTUT4RYJXUEizCfYB7RRHLgeEWfP");
+
+        let payload2 = PaymentRequest::try_from(&mob_url).unwrap();
+
+        assert_eq!(payload, payload2);
+    }
+
+    // Test an example request payload being parsed to mob url
+    #[test]
+    fn example_fog_payload() {
+        let acct = AccountKey::new_with_fog(
+            &RistrettoPrivate::try_from(&[0u8; 32]).unwrap(),
+            &RistrettoPrivate::try_from(&[1u8; 32]).unwrap(),
+            "fog://fog.mobilecoin.com".to_string(),
+            0.to_string(),
+            b"deadbeef".to_vec(),
+        );
+
+        let addr = acct.default_subaddress();
+
+        let mut payload = PaymentRequest::from(&addr);
+
+        payload.amount = Some(777);
+        payload.memo = Some("2 baby goats".to_owned());
+
+        let mob_url = MobUrl::try_from(&payload)
+            .map_err(|err| {
+                panic!("Error when decoding payload {:?}: {}", payload, err);
+            })
+            .unwrap();
+
+        assert_eq!(mob_url.as_ref(), "mob://fog.mobilecoin.com/oGbA6juTWhUdfL6qNMocAGN96wNiZpZegP0TUjKXHEM-GYmM50bLJVeL6NgftIumjt8nwYw7MjEnQT7hCw9bVUgh?a=777&m=2+baby+goats&s=CQkJCfSo#0");
+
+        let payload2 = PaymentRequest::try_from(&mob_url).unwrap();
+
+        assert_eq!(payload, payload2);
+    }
+
+    // Test an example unicode request payload being parsed to mob url
+    #[test]
+    fn roundtrip_example_fog_payload_unicode() {
```

Comment:
what do the urls look like when the memo includes unicode?

Reply:
they look like well-formed URLs, the encoding is done per the URL spec